zerver/actions/message_flags.py

```diff
@@ -289,23 +289,25 @@ def do_update_message_flags(
             user_profile=user_profile, message_id__in=messages
         )
 
-        um_message_ids = {um.message_id for um in query}
-        if flag == "read" and operation == "add":
+        um_flags = {um.message_id: int(um.flags) for um in query}
+
+        if (operation == "add" and ~DEFAULT_HISTORICAL_FLAGS & flagattr != 0) or (
+            operation == "remove" and DEFAULT_HISTORICAL_FLAGS & flagattr != 0
+        ):
             # When marking messages as read, creating "historical"
             # UserMessage rows would be a waste of storage, because
             # `flags.read | flags.historical` is exactly the flags we
             # simulate when processing a message for which a user has
             # access but no UserMessage row.
-            messages = [message_id for message_id in messages if message_id in um_message_ids]
-        else:
+            #
             # Users can mutate flags for messages that don't have a
             # UserMessage yet.  Validate that the user is even allowed
             # to access these message_ids; if so, we will create
             # "historical" UserMessage rows for the messages in question.
             #
             # See create_historical_user_messages for a more detailed
             # explanation.
```

Comment:
The comments here probably could benefit from some reorganization; but it's not terrible.